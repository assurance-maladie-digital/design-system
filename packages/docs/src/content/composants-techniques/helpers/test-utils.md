---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
