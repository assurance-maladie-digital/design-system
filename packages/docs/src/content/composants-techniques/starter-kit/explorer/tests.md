Le dossier `tests` contient les tests d’interface et les tests unitaires globaux de l’application.

<doc-alert type="warning">
Il n’est pas recommandé de renommer ce dossier.
</doc-alert>
