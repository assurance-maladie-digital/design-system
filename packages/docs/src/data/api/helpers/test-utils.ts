import { Api } from '~/types';

export const api: Api = {
	applyGlobalConfig: {
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installGlobalPlugins: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installRouter: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	createRouter: {
		arguments: [
			{
				name: 'options',
				description: 'Router options',
				type: 'RouterOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'VueRouter',
			},
		],
	},
	createStore: {
		arguments: [
			{
				name: 'options',
				description: 'Store options',
				type: 'StoreOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'Store',
			},
		],
	},
	createVuetifyInstance: {
		arguments: [
			{
				name: 'preset',
				description: 'Vuetify preset',
				type: 'Partial<UserVuetifyPreset>',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'Vuetify',
			},
		],
	},
	mockVFormRef: {
		arguments: [
			{
				name: 'isValid',
				description: 'Is valid',
				type: 'boolean',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	mockTranslations: {
		arguments: [
			{
				name: 'translations',
				description: 'Translations',
				type: 'IndexedObject',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
};
