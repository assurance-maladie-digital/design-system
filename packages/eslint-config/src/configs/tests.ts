import { type FlatESLintConfigItem } from 'eslint-define-config';
import { pluginVitest } from '../plugins';
import { GLOB_TESTS } from '../globs';
import { isInEditor } from '../env';

export const tests: FlatESLintConfigItem[] = [
	{
		plugins: {
			tests: pluginVitest,
		},
	},
	{
		files: GLOB_TESTS,
		rules: {
			'tests/consistent-test-it': [
				'error',
				{
					fn: 'it',
					withinDescribe: 'it',
				},
			],
			'tests/expect-expect': 'error',
			'tests/no-alias-methods': 'error',
			'tests/no-commented-out-tests': isInEditor ? 'off' : 'error',
			'tests/no-conditional-expect': 'error',
			'tests/no-disabled-tests': 'error',
			'tests/no-done-callback': 'error',
			'tests/no-duplicate-hooks': 'error',
			'tests/no-focused-tests': isInEditor ? 'off' : 'error',
			'tests/no-identical-title': 'error',
			'tests/no-interpolation-in-snapshots': 'error',
			'tests/no-standalone-expect': 'error',
			'tests/no-test-return-statement': 'error',
			'tests/prefer-comparison-matcher': 'error',
			'tests/prefer-equality-matcher': 'error',
			'tests/prefer-hooks-in-order': 'error',
			'tests/prefer-hooks-on-top': 'error',
			'tests/prefer-mock-promise-shorthand': 'error',
			'tests/prefer-spy-on': 'error',
			'tests/prefer-to-be-falsy': 'error',
			'tests/prefer-to-be-object': 'error',
			'tests/prefer-to-be-truthy': 'error',
			'tests/prefer-to-be': 'error',
			'tests/prefer-to-contain': 'error',
			'tests/prefer-to-have-length': 'error',
			'tests/prefer-todo': 'error',
			'tests/require-hook': 'error',
			'tests/require-top-level-describe': 'error',
			'tests/valid-describe-callback': 'error',
			'tests/valid-expect': 'error',
			'tests/valid-title': 'error',
		},
	},
];
