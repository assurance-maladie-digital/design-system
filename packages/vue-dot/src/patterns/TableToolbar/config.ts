export const config = {
	toolbar: {
		flat: true,
		color: 'white',
		height: 'auto',
		minHeight: '56px',
		class: 'd-flex'
	},
	addBtn: {
		outlined: true,
		color: 'primary',
		class: 'my-1 px-2 px-md-4',
		minWidth: '44px'
	},
	addIconLabel: {
		class: 'mr-1'
	},
	textField: {
		clearable: true,
		singleLine: true,
		hideDetails: true
	}
};
