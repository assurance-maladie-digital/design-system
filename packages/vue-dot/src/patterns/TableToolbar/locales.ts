export const locales = {
    rowText: 'lignes'
};