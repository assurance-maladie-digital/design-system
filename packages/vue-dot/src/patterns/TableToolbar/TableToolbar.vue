<template>
	<VToolbar
		v-bind="options.toolbar"
		class="vd-table-toolbar"
	>
		<p
			v-if="showRowsNumber"
			class="mb-0 font-weight-bold mr-4"
		>
			{{ computedNbRows }} {{ computedRowsText }}
		</p>

		<VSpacer />

		<slot name="search-left" />

		<VTextField
			v-bind="options.textField"
			:value="search"
			:disabled="loading"
			:append-icon="searchIcon"
			:label="searchLabel"
			:class="textFieldClasses"
			@input="$emit('search', $event)"
		/>

		<VBtn
			v-if="showAddBtn"
			v-bind="options.addBtn"
			:disabled="loading"
			@click="$emit('click')"
		>
			<VIcon v-bind="options.addIcon">
				{{ addIcon }}
			</VIcon>

			<span
				v-show="$vuetify.breakpoint.mdAndUp"
				v-bind="options.addIconLabel"
			>
				{{ addBtnLabel }}
			</span>
		</VBtn>

		<slot name="search-right" />
	</VToolbar>
</template>

<script lang="ts">
	import Vue from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { mdiMagnify, mdiPlus } from '@mdi/js';

	import { config } from './config';
	import { locales } from './locales';

	import { customizable } from '../../mixins/customizable';

	const Props = Vue.extend({
		props: {
			nbTotal: {
				type: Number,
				required: true
			},
			nbFiltered: {
				type: Number,
				default: undefined
			},
			search: {
				type: String,
				default: undefined
			},
			searchLabel: {
				type: String,
				default: locales.search
			},
			rowText: {
				type: String,
				default: locales.defaultRowText
			},
			showAddBtn: {
				type: Boolean,
				default: false
			},
			addBtnLabel: {
				type: String,
				default: locales.addBtnLabel
			},
			loading: {
				type: Boolean,
				default: false
			}
		}
	});

	const MixinsDeclaration = mixins(Props, customizable(config));

	@Component({
		model: {
			prop: 'search',
			event: 'search'
		}
	})
	export default class TableToolbar extends MixinsDeclaration {
		searchIcon = mdiMagnify;
		addIcon = mdiPlus;

		get showRowsNumber(): boolean {
			return Boolean(this.nbTotal);
		}

		get computedNbRows(): string {
			return `${this.nbFiltered !== undefined ? (this.nbFiltered + '/') : ''}${this.nbTotal}`;
		}

		get computedRowsText(): string {
			const plural = this.nbTotal > 1;

			return locales.rowText(this.rowText, plural);
		}

		get textFieldClasses(): string {
			let fieldClass = this.$vuetify.breakpoint.xs ? 'vd-form-input-s' : 'vd-form-input';

			if (this.showAddBtn) {
				fieldClass += ' mr-6';
			}

			return `${fieldClass} flex-grow-0`;
		}
	}
</script>

<style lang="scss" scoped>
	.vd-table-toolbar {
		z-index: 1; // Display content above the table on mobile
		contain: none; // Allow fixed elements to be displayed properly

		:deep(.v-toolbar__content) {
			width: 100%;
			flex-wrap: wrap;
		}
	}
</style>
