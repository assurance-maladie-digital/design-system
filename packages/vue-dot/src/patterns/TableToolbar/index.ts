import TableToolbar from './TableToolbar.vue';

export default TableToolbar;
