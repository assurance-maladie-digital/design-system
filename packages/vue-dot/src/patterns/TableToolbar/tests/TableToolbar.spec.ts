import Vue from 'vue';
import { Wrapper } from '@vue/test-utils';

import { mountComponent } from '@/tests';

import TableToolbar from '../';

let wrapper: Wrapper<Vue>;

describe('TableToolbar', () => {
	it('renders correctly', () => {
		wrapper = mountComponent(TableToolbar, {
			propsData: {
				nbFiltered: 1,
				nbTotal: 2,
				showAddBtn: true
			}
		});

		expect(wrapper).toMatchSnapshot();
	});

	it('renders correctly when loading', () => {
		wrapper = mountComponent(TableToolbar, {
			propsData: {
				loading: true,
				nbFiltered: 0,
				nbTotal: 1
			}
		}, true);

		expect(wrapper).toMatchSnapshot();
	});

	it('renders correctly with content slot', () => {
		wrapper = mountComponent(TableToolbar, {
			propsData: {
				nbFiltered: 0,
				nbTotal: 1
			},
			slots: {
				searchLeft: '<p>search-left</p>'
			}
		}, true);

		expect(wrapper).toMatchSnapshot();
	});

	it('renders correctly with no items', () => {
		wrapper = mountComponent(TableToolbar, {
			propsData: {
				nbFiltered: 0,
				nbTotal: 0
			}
		});

		expect(wrapper).toMatchSnapshot();
	});
});
