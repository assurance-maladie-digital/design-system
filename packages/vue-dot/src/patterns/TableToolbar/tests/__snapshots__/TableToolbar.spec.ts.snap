// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableToolbar renders correctly 1`] = `
<vtoolbar-stub color="white" height="auto" minheight="56px" tag="header" extensionheight="48" flat="true" src="" class="vd-table-toolbar d-flex">
  <p class="mb-0 font-weight-bold mr-4">
    1/2 lignes
  </p>
  <vspacer-stub tag="div"></vspacer-stub>
  <vtextfield-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" appendicon="M9.5,3A6.5,6.5 0 0,1 16,9.5C16,11.11 15.41,12.59 14.44,13.73L14.71,14H15.5L20.5,19L19,20.5L14,15.5V14.71L13.73,14.44C12.59,15.41 11.11,16 9.5,16A6.5,6.5 0 0,1 3,9.5A6.5,6.5 0 0,1 9.5,3M9.5,5C7,5 5,7 5,9.5C5,12 7,14 9.5,14C12,14 14,12 14,9.5C14,7 12,5 9.5,5Z" backgroundcolor="" hidedetails="true" label="Rechercher" loaderheight="2" clearable="true" clearicon="$clear" singleline="true" type="text" class="vd-form-input mr-6 flex-grow-0"></vtextfield-stub>
  <vbtn-stub color="primary" minwidth="44px" outlined="true" tag="button" activeclass="" type="button" class="my-1 px-2 px-md-4">
    <vicon-stub>
      M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z
    </vicon-stub> <span class="mr-1">
			Ajouter
		</span>
  </vbtn-stub>
</vtoolbar-stub>
`;

exports[`TableToolbar renders correctly when loading 1`] = `
<header class="vd-table-toolbar v-sheet theme--light v-toolbar v-toolbar--flat white d-flex" style="min-height: 56px;">
  <div class="v-toolbar__content">
    <p class="mb-0 font-weight-bold mr-4">
      0/1 ligne
    </p>
    <div class="spacer"></div>
    <div class="v-input v-input--hide-details v-input--is-disabled theme--light v-text-field v-text-field--single-line vd-form-input flex-grow-0">
      <div class="v-input__control">
        <div class="v-input__slot">
          <div class="v-text-field__slot"><label for="input-8" class="v-label v-label--is-disabled theme--light" style="left: 0px; position: absolute;">Rechercher</label><input disabled="disabled" id="input-8" type="text"></div>
          <div class="v-input__append-inner">
            <div></div>
          </div>
          <div class="v-input__append-inner">
            <div class="v-input__icon v-input__icon--append"><span aria-hidden="true" class="v-icon notranslate v-icon--disabled theme--light"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" role="img" aria-hidden="true" class="v-icon__svg"><path d="M9.5,3A6.5,6.5 0 0,1 16,9.5C16,11.11 15.41,12.59 14.44,13.73L14.71,14H15.5L20.5,19L19,20.5L14,15.5V14.71L13.73,14.44C12.59,15.41 11.11,16 9.5,16A6.5,6.5 0 0,1 3,9.5A6.5,6.5 0 0,1 9.5,3M9.5,5C7,5 5,7 5,9.5C5,12 7,14 9.5,14C12,14 14,12 14,9.5C14,7 12,5 9.5,5Z"></path></svg></span></div>
          </div>
        </div>
      </div>
    </div>
    <!---->
  </div>
</header>
`;

exports[`TableToolbar renders correctly with content slot 1`] = `
<header class="vd-table-toolbar v-sheet theme--light v-toolbar v-toolbar--flat white d-flex" style="min-height: 56px;">
  <div class="v-toolbar__content">
    <p class="mb-0 font-weight-bold mr-4">
      0/1 ligne
    </p>
    <div class="spacer"></div>
    <div class="v-input v-input--hide-details theme--light v-text-field v-text-field--single-line vd-form-input flex-grow-0">
      <div class="v-input__control">
        <div class="v-input__slot">
          <div class="v-text-field__slot"><label for="input-13" class="v-label theme--light" style="left: 0px; position: absolute;">Rechercher</label><input id="input-13" type="text"></div>
          <div class="v-input__append-inner">
            <div></div>
          </div>
          <div class="v-input__append-inner">
            <div class="v-input__icon v-input__icon--append"><span aria-hidden="true" class="v-icon notranslate theme--light"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" role="img" aria-hidden="true" class="v-icon__svg"><path d="M9.5,3A6.5,6.5 0 0,1 16,9.5C16,11.11 15.41,12.59 14.44,13.73L14.71,14H15.5L20.5,19L19,20.5L14,15.5V14.71L13.73,14.44C12.59,15.41 11.11,16 9.5,16A6.5,6.5 0 0,1 3,9.5A6.5,6.5 0 0,1 9.5,3M9.5,5C7,5 5,7 5,9.5C5,12 7,14 9.5,14C12,14 14,12 14,9.5C14,7 12,5 9.5,5Z"></path></svg></span></div>
          </div>
        </div>
      </div>
    </div>
    <!---->
  </div>
</header>
`;

exports[`TableToolbar renders correctly with no items 1`] = `
<vtoolbar-stub color="white" height="auto" minheight="56px" tag="header" extensionheight="48" flat="true" src="" class="vd-table-toolbar d-flex">
  <!---->
  <vspacer-stub tag="div"></vspacer-stub>
  <vtextfield-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" appendicon="M9.5,3A6.5,6.5 0 0,1 16,9.5C16,11.11 15.41,12.59 14.44,13.73L14.71,14H15.5L20.5,19L19,20.5L14,15.5V14.71L13.73,14.44C12.59,15.41 11.11,16 9.5,16A6.5,6.5 0 0,1 3,9.5A6.5,6.5 0 0,1 9.5,3M9.5,5C7,5 5,7 5,9.5C5,12 7,14 9.5,14C12,14 14,12 14,9.5C14,7 12,5 9.5,5Z" backgroundcolor="" hidedetails="true" label="Rechercher" loaderheight="2" clearable="true" clearicon="$clear" singleline="true" type="text" class="vd-form-input flex-grow-0"></vtextfield-stub>
  <!---->
</vtoolbar-stub>
`;
